### PR TITLE
Fixes #12140 - unreported scoped search errors on 500 field

### PR DESCRIPTION
**Pull request by johnpmitsch**
This will report scoped search errors in a notification in the UI

### CODE REVIEW

**Comment by johnpmitsch** [comment]
These bastion changes are needed https://github.com/Katello/bastion/pull/97

File: app/controllers/katello/api/v2/api_controller.rb [hunk 1]
```diff
@@ -76,18 +76,27 @@ def scoped_search(query, default_sort_by, default_sort_order, options = {})
       else
         query = query.paginate(paginate_options)
       end
+      page = params[:page] || 1
+      per_page = params[:per_page] || ::Setting::General.entries_per_page
 
+      scoped_search_results(query, sub_total, total, page, per_page)
+    rescue Exception => error
+      return scoped_search_results(query, sub_total, total, page, per_page, error.message)
```
Comment (ehelms) [inline app/controllers/katello/api/v2/api_controller.rb]: Is the explicit return required?

**Comment by ehelms** [comment]
Thoughts on a unit test for this case?

File: app/controllers/katello/api/v2/api_controller.rb [hunk 2]
```diff
@@ -76,18 +76,27 @@ def scoped_search(query, default_sort_by, default_sort_order, options = {})
       else
         query = query.paginate(paginate_options)
       end
+      page = params[:page] || 1
+      per_page = params[:per_page] || ::Setting::General.entries_per_page
 
+      scoped_search_results(query, sub_total, total, page, per_page)
+    rescue Exception => error
```
Comment (daviddavis) [inline app/controllers/katello/api/v2/api_controller.rb]: [Rescue StandardError not Exception](https://robots.thoughtbot.com/rescue-standarderror-not-exception).

**Comment by daviddavis** [comment]
I tested this out and when I caused the database to timeout when performing a scoped search, I got back a 200 response (instead of a 500 response) and there was no backtrace in my logs either. Is that the expected behavior?

**Comment by johnpmitsch** [comment]
@ehelms I'm having a hard time writing tests for the scoped_search method. I am trying to do something like 
``` response = @controller.scoped_search(ActivationKey.all, :name, :asc, :resource_class => Katello::ActivationKey) ``` 
but unsure of how to pass a search string to the scoped_search function, which comes from this [search_options variable](https://git.io/v2rqU)

Doing something like ```response = ActivationKey.search_for("foo = bar")``` doesn't hit the scoped search method and returns ```ScopedSearch::QueryNotSupported: Field 'foo' not recognized for searching!```

**Comment by johnpmitsch** [comment]
updated to rescue from ScopedSearch::Exception which scoped search [errors inherit from](https://git.io/v2rO8)

**Comment by johnpmitsch** [comment]
Changed the rescue again to ScopedSearch::QueryNotSupported as the only other scoped search error [should not be rescued](https://github.com/wvanbergen/scoped_search/blob/94201712f20c244195001cf14b53b494eec9b42b/lib/scoped_search.rb#L48-L49)  (thanks to @daviddavis for finding that)

**Comment by ehelms** [comment]
ACK

**Comment by daviddavis** [comment]
ACK

**Comment by johnpmitsch** [comment]
@ehelms should I wait on an ack and merge of https://github.com/Katello/bastion/pull/97, and a new release of bastion before merging this?

**Comment by ehelms** [comment]
As I understand it, this will improve the API response without breaking the UI or CLI as is? And the bastion change will allow showing this to the user?

**Comment by johnpmitsch** [comment]
@ehelms that is correct, I shall merge :)

**Pull request merged**